File: calc/rustfwcalc/tests/calculator_model_test.rs
use rustcalc::model::{Action, CalculatorModel, Operation};

#[test]
fn test_initial_state() {
    let model = CalculatorModel::new();
    assert_eq!(model.current_input(), "0");
    assert_eq!(model.expression_text(), " ");
    assert!(!model.is_error());
}

#[test]
fn test_digit_input() {
    let mut model = CalculatorModel::new();
    model.dispatch(Action::Digit('1'));
    model.dispatch(Action::Digit('2'));
    model.dispatch(Action::Digit('3'));
    assert_eq!(model.current_input(), "123");
}

#[test]
fn test_decimal_input() {
    let mut model = CalculatorModel::new();
    model.dispatch(Action::Digit('5'));
    model.dispatch(Action::Decimal);
    model.dispatch(Action::Digit('2'));
    assert_eq!(model.current_input(), "5.2");

    model.dispatch(Action::Decimal);
    assert_eq!(model.current_input(), "5.2");
}

#[test]
fn test_addition() {
    let mut model = CalculatorModel::new();
    model.dispatch(Action::Digit('5'));
    model.dispatch(Action::SetOperator(Operation::Add));
    assert_eq!(model.expression_text(), "5 ＋");

    model.dispatch(Action::Digit('3'));
    model.dispatch(Action::Equals);

    assert_eq!(model.current_input(), "8");
    assert_eq!(model.expression_text(), "5 ＋ 3 =");
}

#[test]
fn test_division_by_zero() {
    let mut model = CalculatorModel::new();
    model.dispatch(Action::Digit('9'));
    model.dispatch(Action::SetOperator(Operation::Divide));
    model.dispatch(Action::Digit('0'));
    model.dispatch(Action::Equals);

    assert!(model.is_error());
    assert_eq!(model.current_input(), "Error: Division by zero");
}

#[test]
fn test_chained_operations() {
    let mut model = CalculatorModel::new();
    model.dispatch(Action::Digit('1'));
    model.dispatch(Action::Digit('0'));
    model.dispatch(Action::SetOperator(Operation::Subtract));
    model.dispatch(Action::Digit('4'));
    model.dispatch(Action::SetOperator(Operation::Multiply));
    assert_eq!(model.current_input(), "6");
    assert_eq!(model.expression_text(), "6 ×");

    model.dispatch(Action::Digit('5'));
    model.dispatch(Action::Equals);
    assert_eq!(model.current_input(), "30");
}

#[test]
fn test_backspace() {
    let mut model = CalculatorModel::new();
    model.dispatch(Action::Digit('1'));
    model.dispatch(Action::Digit('2'));
    model.dispatch(Action::Digit('3'));
    model.dispatch(Action::Backspace);
    assert_eq!(model.current_input(), "12");

    model.dispatch(Action::Backspace);
    model.dispatch(Action::Backspace);
    assert_eq!(model.current_input(), "0");
}

#[test]
fn test_toggle_sign() {
    let mut model = CalculatorModel::new();
    model.dispatch(Action::Digit('7'));
    model.dispatch(Action::ToggleSign);
    assert_eq!(model.current_input(), "-7");

    model.dispatch(Action::ToggleSign);
    assert_eq!(model.current_input(), "7");
}

#[test]
fn test_reset() {
    let mut model = CalculatorModel::new();
    model.dispatch(Action::Digit('8'));
    model.dispatch(Action::SetOperator(Operation::Add));
    model.dispatch(Action::Digit('4'));
    model.dispatch(Action::Clear);

    assert_eq!(model.current_input(), "0");
    assert_eq!(model.expression_text(), " ");
    assert_eq!(model.current_operator(), None);
}

File: calc/javacalc/src/test/java/com/javacalc/CalculatorModelTest.java
package com.javacalc;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class CalculatorModelTest {

    private CalculatorModel model;

    @BeforeEach
    void setUp() {
        model = new CalculatorModel();
    }

    @Test
    void testInitialState() {
        assertEquals("0", model.getCurrentInput());
        assertEquals(" ", model.getExpressionText());
        assertFalse(model.isError());
    }

    @Test
    void testDigitInput() {
        model.enterDigit("1");
        model.enterDigit("2");
        model.enterDigit("3");
        assertEquals("123", model.getCurrentInput());
    }

    @Test
    void testDecimalInput() {
        model.enterDigit("5");
        model.enterDecimal();
        model.enterDigit("2");
        assertEquals("5.2", model.getCurrentInput());

        // Repeated decimal points should be ignored
        model.enterDecimal();
        assertEquals("5.2", model.getCurrentInput());
    }

    @Test
    void testAddition() {
        model.enterDigit("5");
        model.setOperator(Operation.ADD);
        assertEquals("5 ＋", model.getExpressionText());

        model.enterDigit("3");
        model.calculateEquals();

        assertEquals("8", model.getCurrentInput());
        assertEquals("5 ＋ 3 =", model.getExpressionText());
    }

    @Test
    void testDivisionByZero() {
        model.enterDigit("9");
        model.setOperator(Operation.DIVIDE);
        model.enterDigit("0");
        model.calculateEquals();

        assertTrue(model.isError());
        assertEquals("Error: Division by zero", model.getCurrentInput());
    }

    @Test
    void testChainedOperations() {
        model.enterDigit("1");
        model.enterDigit("0"); // 10
        model.setOperator(Operation.SUBTRACT);
        model.enterDigit("4"); // - 4
        model.setOperator(Operation.MULTIPLY); // evaluates to 6, then pending *
        assertEquals("6", model.getCurrentInput());
        assertEquals("6 ×", model.getExpressionText());

        model.enterDigit("5");
        model.calculateEquals();
        assertEquals("30", model.getCurrentInput());
    }

    @Test
    void testBackspace() {
        model.enterDigit("1");
        model.enterDigit("2");
        model.enterDigit("3");
        model.backspace();
        assertEquals("12", model.getCurrentInput());

        model.backspace();
        model.backspace();
        assertEquals("0", model.getCurrentInput());
    }

    @Test
    void testToggleSign() {
        model.enterDigit("7");
        model.toggleSign();
        assertEquals("-7", model.getCurrentInput());

        model.toggleSign();
        assertEquals("7", model.getCurrentInput());
    }

    @Test
    void testReset() {
        model.enterDigit("8");
        model.setOperator(Operation.ADD);
        model.enterDigit("4");
        model.reset();

        assertEquals("0", model.getCurrentInput());
        assertEquals(" ", model.getExpressionText());
        assertNull(model.getCurrentOperator());
    }
}
